app/src/main/java/com/github/gotify/service/WebSocketService.java

```diff
@@ -300,4 +336,44 @@ public void showNotificationGroup(long priority) {
                 (NotificationManager) this.getSystemService(Context.NOTIFICATION_SERVICE);
         notificationManager.notify(-5, b.build());
     }
+
+    private Bitmap getIcon(Integer appid) {
+
+        if(appid==-1){
+            return BitmapFactory.decodeResource(getResources(), R.drawable.gotify);
+        }
+
+        Bitmap icon;
+        RequestCreator request;
+        try {
+            request= cache.load(Utils.resolveAbsoluteUrl(settings.url() + "/", appIdMap.get(appid)));
+            icon = request.get();
+        } catch (IOException e) {
+            icon = BitmapFactory.decodeResource(getResources(), R.drawable.gotify);
+        }
+
+        return icon;
+    }
+
+    private Picasso makePicasso() {
+        Cache picassoCache = new Cache(new File(getCacheDir(), "picasso-cache"), MessagesActivity.PICASSO_CACHE_SIZE);
+        OkHttpClient.Builder builder = new OkHttpClient.Builder();
+        builder.cache(picassoCache);
+        CertUtils.applySslSettings(builder, settings.sslSettings());
+        OkHttp3Downloader downloader = new OkHttp3Downloader(builder.build());
+        return new Picasso.Builder(this).downloader(downloader).build();
+    }
+
+    private void updateAppIds(){
+        ApiClient client = ClientFactory.clientToken(settings.url(), settings.sslSettings(), settings.token());
+
+        List<Application> applications = null;
+        try {
+            applications = client.createService(ApplicationApi.class).getApps().execute().body();
+            appIdMap = MessageImageCombiner.appIdToImage(applications);
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
```

Comment:
`#enqueue` can be used to asynchronously execute the http request. This removes the need for the `new Thread()..` inside `com.github.gotify.service.WebSocketService#startPushService` 
```suggestion
    private void updateAppIds(){
        ClientFactory.clientToken(settings.url(), settings.sslSettings(), settings.token())
                .createService(ApplicationApi.class)
                .getApps()
                .enqueue(Callback.call(
                                (apps) -> {
                                    appIdMap.clear();
                                    appIdMap.putAll(MessageImageCombiner.appIdToImage(apps));
                                },
                                (t) -> {
                                    appIdMap.clear();
                                }));
```
Using Callback.call automatically logs errors.